servers/src/main/java/tachyon/master/next/filesystem/meta/InodeTree.java

```diff
@@ -253,57 +251,87 @@ public void setPinned(Inode inode, boolean pinned) {
+    if (pathComponents == null) {
+      throw new InvalidPathException("passed-in pathComponents is null");
+    } else if (pathComponents.length == 0) {
+      throw new InvalidPathException("passed-in pathComponents is empty");
+    } else if (pathComponents.length == 1) {
       if (pathComponents[0].equals("")) {
-        return new Pair<Inode, Integer>(mRoot, -1);
+        return TraverseResult.createFoundResult(mRoot);
       } else {
-        final String msg = "File name starts with " + pathComponents[0];
-        LOG.info("InvalidPathException: " + msg);
-        throw new InvalidPathException(msg);
+        throw new InvalidPathException("File name starts with " + pathComponents[0]);
       }
     }
 
-    Pair<Inode, Integer> ret = new Pair<Inode, Integer>(mRoot, -1);
+    Inode current = mRoot;
 
-    for (int k = 1; k < pathComponents.length; k ++) {
-      Inode next = ((InodeDirectory) ret.getFirst()).getChild(pathComponents[k]);
+    for (int i = 1; i < pathComponents.length; i ++) {
+      Inode next = ((InodeDirectory) current).getChild(pathComponents[i]);
       if (next == null) {
-        // The user might want to create the nonexistent directories, so we leave ret.getFirst()
-        // as the last Inode taken. We set nonexistentInd to k, to indicate that the kth path
+        // The user might want to create the nonexistent directories, so return the current inode
+        // directory
+        // as the last Inode taken. Also set nonexistentInd to k, to indicate that the kth path
         // component was the first one that couldn't be found.
-        ret.setSecond(k);
-        break;
-      }
-      ret.setFirst(next);
-      if (!ret.getFirst().isDirectory()) {
+        return TraverseResult.createNotFoundResult(current, i);
+      } else if (next.isFile()) {
         // The inode can't have any children. If this is the last path component, we're good.
         // Otherwise, we can't traverse further, so we clean up and throw an exception.
-        if (k == pathComponents.length - 1) {
-          break;
+        if (i == pathComponents.length - 1) {
+          return TraverseResult.createFoundResult(next);
         } else {
-          final String msg =
-              "Traversal failed. Component " + k + "(" + ret.getFirst().getName() + ") is a file";
-          LOG.info("InvalidPathException: " + msg);
-          throw new InvalidPathException(msg);
+          throw new InvalidPathException(
+              "Traversal failed. Component " + i + "(" + next.getName() + ") is a file");
         }
+      } else {
+        // next is a directory and keep navigating
+        current = next;
       }
     }
-    return ret;
+    return TraverseResult.createFoundResult(current);
   }
 
+  private static final class TraverseResult {
+    private final boolean mFound;
+    /**
+     * when the path is not found in a traversal, the index of the last navigated step in the path
```

Comment:
Is this index the first one NOT found, or the last one found? I think the comments are not clear.